processing/src/main/java/org/apache/druid/frame/field/DoubleFieldWriter.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.field;
+
+import org.apache.datasketches.memory.WritableMemory;
+import org.apache.druid.segment.BaseDoubleColumnValueSelector;
+
+/**
+ * Wraps a {@link BaseDoubleColumnValueSelector} and writes field values.
+ *
+ * Values are transformed such that they are comparable as bytes; see {@link #transform} and {@link #detransform}.
+ * Values are preceded by a null byte that is either 0x00 (null) or 0x01 (not null). This ensures that nulls sort
+ * earlier than nonnulls.
+ */
+public class DoubleFieldWriter implements FieldWriter
+{
+  public static final int SIZE = Double.BYTES + Byte.BYTES;
+  public static final byte NULL_BYTE = 0x00;
+  public static final byte NOT_NULL_BYTE = 0x01;
+
+  private final BaseDoubleColumnValueSelector selector;
+
+  public DoubleFieldWriter(final BaseDoubleColumnValueSelector selector)
+  {
+    this.selector = selector;
+  }
+
+  @Override
+  public long writeTo(final WritableMemory memory, final long position, final long maxSize)
```

Comment:
Example of the above comment. Here, the caller has to keep track of the offset for each field, and update that offset using the returned length. That's quite a bit of accounting if we have, say, 100 fields!

This class appears to be static: it accepts the memory, the position and a value (from a selector?) but yet another thing keeps track of offsets.

Might be cleaner if this object is created per field and tracks offsets internally. If we assume sequential writes (that is, row-by-row), then the write is only moving forward: works for fixed-width and variable-width values. If we allow random writes (not sure how we'd do that), can we explain the logic?


Did we consider using a row offset so we can do the math for fixed fields as above? For variable-width fields, is there

Reply:
RowBasedFrameWriter is the main place this is used; it uses the return value of these field-writer functions (bytes written) to increase a single `bytesWritten` pointer that it tracks. It seems clean to me but please let me know if I'm missing something.